tests/unit/s2n_self_talk_state_machine_test.c

```diff
@@ -0,0 +1,189 @@
+
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS12);
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(client_conn), S2N_TLS12);
+
+            /* Alter which state machine the connection is using */
+            server_conn->actual_protocol_version = S2N_TLS13;
+            client_conn->actual_protocol_version = S2N_TLS13;
+
+            EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn), S2N_ERR_STATE_MACHINE_SWITCH);
+        }
+
+        /* An early data handshake cannot change to the TLS12 state machine */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+            EXPECT_SUCCESS(s2n_connection_set_blinding(server_conn, S2N_SELF_SERVICE_BLINDING));
+
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            EXPECT_SUCCESS(s2n_connection_set_blinding(client_conn, S2N_SELF_SERVICE_BLINDING));
+
+            /* Create nonblocking pipes */
+            struct s2n_test_io_pair io_pair;
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+            const uint8_t test_data[] = "hello world";
+            DEFER_CLEANUP(struct s2n_psk *test_psk = s2n_external_psk_new(), s2n_psk_free);
+            EXPECT_SUCCESS(s2n_psk_set_identity(test_psk, test_data, sizeof(test_data)));
+            EXPECT_SUCCESS(s2n_psk_set_secret(test_psk, test_data, sizeof(test_data)));
+            EXPECT_SUCCESS(s2n_psk_configure_early_data(test_psk, UINT32_MAX, 0x13, 0x01));
+
+            EXPECT_SUCCESS(s2n_connection_append_psk(client_conn, test_psk));
+            EXPECT_SUCCESS(s2n_connection_append_psk(server_conn, test_psk));
+
+            uint8_t actual_payload[BUFFER_SIZE] = { 0 };
+            s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+            ssize_t data_size = 0;
+
+            /* Send and receive some early data */
+            EXPECT_FAILURE_WITH_ERRNO(s2n_send_early_data(client_conn, test_data, sizeof(test_data), &data_size, &blocked), S2N_ERR_IO_BLOCKED);
+            EXPECT_FAILURE_WITH_ERRNO(s2n_recv_early_data(server_conn, actual_payload, sizeof(actual_payload), &data_size, &blocked), S2N_ERR_IO_BLOCKED);
+
+            /* Alter which state machine the connection is using */
+            client_conn->actual_protocol_version = S2N_TLS12;
+            server_conn->actual_protocol_version = S2N_TLS12;
+
+            EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn), S2N_ERR_STATE_MACHINE_SWITCH);
```

Comment:
This test is only interesting if the error happens before the ServerHello-- after that it's equivalent to the first test. Can you make an assertion about which message we're on? I think we're actually past the ServerHello here (at least for the server) since s2n_send/recv_early_data try to make progress on the handshake.